### PR TITLE
create entries batch multiple entries into single insert

### PR DESCRIPTION
**Pull request by nopcoder**
- merge all cataloger configuration parameters into params.Catalog

Added

"cataloger.batch_write.insert_size"
"cataloger.cache.enabled"
"cataloger.cache.size"
"cataloger.cache.expiry"
"cataloger.cache.jitter"

### CODE REVIEW

File: catalog/cataloger_create_entries.go [hunk 1]
```diff
@@ -33,9 +36,30 @@ func (c *cataloger) CreateEntries(ctx context.Context, repository, branch string
...
+			sqInsert := psql.Insert("catalog_entries").
+				Columns("branch_id", "path", "physical_address", "checksum", "size", "metadata", "creation_date", "is_expired")
+			j := i + c.CreateEntriesInsertSize
+			if j > len(entries) {
+				j = len(entries)
+			}
+			for _, entry := range entries[i:j] {
+				creationDate := entry.CreationDate
+				if entry.CreationDate.IsZero() {
+					creationDate = time.Now()
```
Comment (arielshaqed) [inline catalog/cataloger_create_entries.go:50]: This is a different clock than on the database.  I think we should use "now" on the database, like we now [do on CreateEntry](https://github.com/treeverse/lakeFS/blob/8b28f3affda313c96afd385a8c58e54e813e6ea3/catalog/cataloger_create_entry.go#L56).

File: catalog/cataloger_create_entries.go [hunk 2]
```diff
@@ -33,9 +36,30 @@ func (c *cataloger) CreateEntries(ctx context.Context, repository, branch string
 		if err != nil {
 			return nil, err
 		}
-		for i := range entries {
-			if _, err := insertEntry(tx, branchID, &entries[i]); err != nil {
-				return nil, fmt.Errorf("entry at %d: %w", i, err)
+		// single insert per batch
+		for i := 0; i < len(entries); i += c.CreateEntriesInsertSize {
```
Comment (arielshaqed) [inline catalog/cataloger_create_entries.go:40]: When there are duplicate create-entries, this behaves unpredictably:
- If a dupe appears inside a batch, we receive an error from INSERT (and I assume a strange error message).
- If a dupe appears between batches, the latest version wins.

I don't particularly care whether it should succeed or fail, but if I had to rank them I'd go with:
1. The behaviour on AWS S3.
2. Latest version wins, always.
3. A decipherable error, always, and _no_ file is created.
4. A decipherable error, always, and _some_ of the files may be created.

In order to do any of these we _will_ have to examine all entries in the batch.

**Review by arielshaqed — COMMENTED** [review]
Thanks, nice and clear!

Not sure I'm a fan of squashing the params out of `cataloger.batchParams`, because I *like* parameter objects.  We could even have a single `config` call that passed the `BatchParams` for a configuration, in line with what we did elsewhere.

Blocker is behaviour on creating the same object multiple times.  Single clock is important but I won't block on it.

**Review by arielshaqed — CHANGES_REQUESTED** [review]
I stated changes but forgot to Request Changes.  A wise man once said (almost):
> You see, you know how to *state* the changes, you just don't know how to *request* the changes. And that's really the most important part of the request: the blocking. Anybody can just state the changes.

File: catalog/cataloger_create_entries.go (hunk 2 above)
Comment (nopcoder) [reply catalog/cataloger_create_entries.go:40]: Going with "Latest version wins"

**Comment by nopcoder** [comment]
catalog params split into parts in code and configuration

File: catalog/cataloger_create_entries.go [hunk 3]
```diff
@@ -23,11 +24,23 @@ func (c *cataloger) CreateEntries(ctx context.Context, repository, branch string
...
+		entriesMap[p] = &entries[i]
+	}
+
+	// prepare a list of entries to insert without duplicates
+	entriesToInsert := make([]*Entry, 0, len(entriesMap))
+	for i := range entries {
+		ent := entriesMap[entries[i].Path]
+		if &entries[i] == ent {
+			entriesToInsert = append(entriesToInsert, ent)
+		}
```
Comment (arielshaqed) [inline catalog/cataloger_create_entries.go:43]: IIUC the goal here is to keep the insert the *last* object for each key in the order of the *first* appearance of each key.  This is a bit odd.  To keep the files in their expected ordering, I would write something like this (untested, of course):
```go
for i := len(entries)-1, j := len(entries); i >= 0; i-- {
	if &entries[i] == entriesMap[entries[i].Path] {
		j--
		entries[j] := entries[i]
	}
}
entries = entries[j:]
```

It *is* a bit more subtle, but it gives a better ordering I think.  If you don't care about the ordering, just iterate over the map.

If you want the last-last ordering above but without editing the slice in-place:
```go
entriesToInsert := make([]*Entry, len(entriesMap))
for i := len(entries)-1, j := len(entries) - 1; i >= 0; i-- {
	if &entries[i] == entriesMap[entries[i].Path] {
		entriesToInsert[j] := entries[i]
		j--
	}
}
```

**Review by arielshaqed — APPROVED** [review]
Niiiice!  Thanks.

File: catalog/cataloger_create_entries.go (hunk 3 above)
Comment (arielshaqed) [reply catalog/cataloger_create_entries.go:43]: F2Fing, the existing code does exactly what I want.  Sorry.

Comment (nopcoder) [reply catalog/cataloger_create_entries.go:43]: like the first loop - I used slice of pointers and keep just the right references in the new slice that we process.

**Pull request merged**